### PR TITLE
Rewrite migrations to not depend on future code changes

### PR DESCRIPTION
**Pull request by daviian**
Backport of #2602

### CODE REVIEW

**Comment by daviian** [comment]
Currently there is an issue in v39 (introduced in this PR) which needs to be fixed before merging

**Comment by lafriks** [comment]
@daviian what issue?

**Comment by daviian** [comment]
@lafriks https://github.com/daviian/gitea/blob/c9430b29a5220e2297aa9b93b70a25062f831509/models/migrations/v39.go#L73
Xorm doesn't return anything here, despite there should be 96 units in my testdata with UnitTypeIssue

File: models/migrations/v37.go [hunk 1]
```diff
@@ -7,24 +7,27 @@ package migrations
...
+		ID       int64 `xorm:"pk autoincr"`
+		FullName string
+	}
+
 	const batchSize = 100
 	for start := 0; ; start += batchSize {
-		users := make([]*models.User, 0, batchSize)
-		if err := x.Limit(start, batchSize).Find(users); err != nil {
+		users := make([]*User, 0, batchSize)
+		if err := x.Limit(batchSize, start).Find(&users); err != nil {
```
Comment (daviian) [inline models/migrations/v37.go:22]: Good catch. Nearly impossible to see

**Comment by strk** [comment]
Any chance to see an automated test for migration from a dump of a Gogs0.99 database ?

**Comment by lafriks** [comment]
@strk not in the scope of this PR but otherwise I agree we would need that but I currently have no idea how to achieve that

**Comment by strk** [comment]
Trusted/untested LGTM - please make sure to manually test it !

**Comment by lafriks** [comment]
@strk me and @daviian have tested it multiple times

**Comment by mrexodia** [comment]
Just tried this on a database with `version=17` and it appears to have worked! (raspberry pi zero w, had to [change swap size](https://www.bitpi.co/2015/02/11/how-to-change-raspberry-pis-swapfile-size-on-rasbian/) to 1024mb to compile)

```
# Make errors about go-bindata go away
go get github.com/jteeuwen/go-bindata
cd $GOPATH/src/github.com/jteeuwen/go-bindata/go-bindata
go build
export PATH=$PATH:$GOPATH/src/github.com/jteeuwen/go-bindata/go-bindata

# Get gitea
go get -d -u code.gitea.io/gitea
cd $GOPATH/src/code.gitea.io/gitea

# Get PR 2604 that fixes migration issues
git fetch origin pull/2604/head:pr-2604
git checkout pr-2604

# Build gitea (will take some time)
TAGS="bindata sqlite" make generate build
```

I kind of followed https://docs.gitea.io/en-us/install-from-source/ and https://docs.gitea.io/en-us/upgrade-from-gogs/ and @lafriks told me to change the `gogs.db` with the following SQL command: `update version set version=14;` which may or may not have contributed to this being successful...

**Comment by daviian** [comment]
@mrexodia Glad to here that 😃 Setting the version to 14 is necessary as otherwise required migrations (in your example 14 to 17) are not applied to the database.

**Comment by Morlinest** [comment]
LGTM, code looks good, trusting manual tests of @lafriks & @daviian

**Comment by lafriks** [comment]
Make LG-TM work again

**Comment by daviian** [comment]
@lunny @lafriks Should I create a backport to 1.2?

**Comment by lafriks** [comment]
@daviian sure, go ahead

**Pull request merged**